doc/developer/design/20231201_catalog_migration_to_persist.md

```diff
@@ -0,0 +1,164 @@
+# Catalog Backend Migration
+
+- Associated: https://github.com/MaterializeInc/materialize/issues/20953
+- Associated: https://github.com/MaterializeInc/materialize/issues/22392
+
+## The Problem
+
+Currently, we persist all users catalog data in the stash. We would like to move all of that data
+to Persist. This will have the following benefits:
+
+- A step towards architecting a shareable catalog that will contribute to use case isolation.
+- Reduced operational overhead related to maintaining the stash. Note, the overhead is not
+  completely removed because the storage controller still uses the stash.
+
+## Success Criteria
+
+All user catalog data is stored in Persist.
+
+## Out of Scope
+
+- Fully shareable differential catalog.
+- Use case isolation.
+- Zero downtime upgrades.
+
+## Solution Proposal
+
+### Config Collection
+
+The config collection is a subset of catalog data that can never be migrated. As a consequence, this
+data can be read before the catalog is fully opened and can be used to bootstrap an environment. The
+config collection has keys of type `String` and values of type `u64`. Booleans can be modeled using
+this collection by storing a `0` for `false` and `1` for `true`. This proposal will add a key called
+`tombstone`, with a boolean value that indicates if a specific catalog backend is retired. A value
+of `false` means that the backend is not retired (i.e. is still in use) and a value of `true` means
+the backend is retired (i.e. is not still in use).
+
+### Epoch Fencing
+
+The catalog durably stores a fencing token called the epoch. All catalog writers and readers also
+store an in-memory epoch. When a new writer or reader connects to the catalog they store the
+previous epoch plus one in memory and persist this new value durably. Every time a reader or write
+performs a read or write, they compare their in-memory token with the persisted token, and if
+there's a difference, the operation fails. As a result, whenever a reader/write increments the epoch
+durably, they are effectively fencing out all previous readers/writers.
+
+### Catalog Open
+
+Opening a catalog involves the following two steps:
+
+1. Increment the epoch.
+2. Either initialize the persisted catalog state if it's uninitialized, or migrate the data format
+   if there are any migrations, otherwise do nothing.
+
+### Migrate From Stash to Persist
+
+Below are the steps to migrate from the stash to persist.
+
+1. Open the stash catalog.
+2. Open the persist catalog.
+3. If the stash tombstone is `true`, then we're done.
+4. Read stash catalog snapshot.
+5. Write catalog snapshot to persist.
```

Comment:
Do we care about the previous state of the persist catalog? Must it be empty? Should be written down.

Reply:
Just pushed an update, we should replace whatever the previous state was with the new snapshot.